### ISSUE TITLE
Generic types extending Record<Uppercase<string>, ...> ignore the Uppercase utility

### ISSUE THREAD

**Issue by sadkebab** [microsoft/TypeScript#60290]
### 🔎 Search Terms

Record<Uppercase<string>
Uppercase

### 🕗 Version & Regression Information

I tested the following snippet on the playground with all versions starting from 4.9.5 and the behaviour was the same.
For all the previous versions down to 4.1 (when Uppercase was introduced) the behaviour was different but the errors seem unrelated.

### ⏯ Playground Link

https://www.typescriptlang.org/play/?ts=5.6.3#code/C4TwDgpgBAqmkCcDGBDAzhAShJB7BAJlALxTZ6EA8ciqGlawCAlgHYDmAfADRQCurANatcAd1acA3AChpAMwFJgzXKyhzcuSgBUoEAB7AIrAmljwIydFhz4CnABQAjFAgBcUbQEooAb2lQUAgQwHwIai4IMgC+sgD0cVDAABbMZmjJuHwANkQpCGJQKGqWBQjSeKyMUBlZuQBiKMzZJOqaDv6BGrgeAESRvdzS0V4yspXVck3ZaADqzCkAyijKaHLMEGak3R0BbT1Q-a6DwzUraeub5rTW5Haj46rVC7P4glv7u4H1API-fQMhiMxhNgFAXm85gtkstVpc0NpNK0dp0oL9-odAac0Oc1hszDRLHQbBQCA9pAkoKJkitwXIqdACKoAORgvgYcFgthFKDsYyWZhIdSKZSqAD88hFKjUoEgBHq7UiHkJVgwd0IPlRwVC4SgkRi8USKWgsugokhFSeXOAjWaH1N8vaqO6AOOQNGQA

### 💻 Code

```ts
type UppercaseRecord = Record<Uppercase<string>, unknown>;

function foo<T extends UppercaseRecord>(bar: T) {
  return bar;
}

// this should throw an error
const shouldFail = foo({
  foo: "bar",
});

const failsWithSatisfies = foo({
  foo: "bar",
} satisfies UppercaseRecord);

const itWorks = foo({
  FOO: "bar",
});

const itWorksWithSatisfiesToo = foo({
  FOO: "bar",
} satisfies UppercaseRecord);

// what if we don't use it in a generic function?
function typedFoo(bar: UppercaseRecord) {
  return bar;
}

const itFails = typedFoo({
  foo: "bar",
});
```

### 🙁 Actual behavior

Passing a record with lowercase properties on foo does not throw an error.

### 🙂 Expected behavior

Passing a record with lowercase properties on foo should throw an error.

### Additional information about the issue

Generic types extending Uppercase<...> seem to work properly.

```ts
function foo<T extends Uppercase<string>>(bar: T) {
  return bar;
}

foo("FOO");

// this throws an error
foo("foo");
```

**Comment by MartinJohns** [comment]
This is working as intended. A record with upper- and lowercase keys does extend a record with only uppercase keys, so your generic type constraint is fulfilled.

**Comment by sadkebab** [comment]
@MartinJohns if I understood what you mean correctly then extending a literal string like Uppercase<"prop"> would produce the same effect since a a Record<string, ...> does extend Record<Uppercase<"prop">> too... but that throws an error.
Isn't this a inconsistency or am I getting something wrong?

**Comment by MartinJohns** [comment]
No, it wouldn't produce the same, since the one is a literal, the other an object. The string `"abc"` does not extend the string `"ABC"`, but the object `{ A: number, b: number }` extends the object type `{ A: number }`. Every property that must exist in `{ A: number }` also exists in `{ A: number, b: number }`.

You seem to mistakingly think that `Record<Uppercase<string>, unknown>` means the object may only have uppercase keys, but that's not the case. It means all uppercase keys must have a value typed `unknown`. You can't prevent the existence of additional keys in TypeScript.

**Comment by sadkebab** [comment]
@MartinJohns Ok I get it now, thank you. How would I restrict a record to having only uppercase string properties then?

**Comment by MartinJohns** [comment]
> How would I restrict a record to having only uppercase string properties then?

You can't, hence I wrote before:

> You can't prevent the existence of additional keys in TypeScript.

See also: https://github.com/microsoft/typescript/wiki/faq#what-is-structural-typing
